pkg/cloud/services/compute/instance.go

```diff
@@ -207,10 +184,46 @@ func (s *Service) createInstanceImpl(eventObject runtime.Object, openStackCluste
 		portList = append(portList, servers.Network{
 			Port: port.ID,
 		})
+
+		if created && openStackMachine != nil {
+			openStackMachine.Status.CreatedPorts = append(openStackMachine.Status.CreatedPorts, port.ID)
+		}
```

Comment:
I very much like the idea of putting ports in the Status. The API change looks good to me.

I think there's a bug in this implementation, though, which suggests how this change lets us simplify this cleanup flow a bit: we're adding created ports to the status here, but we're not removing them if we subsequently delete the ports without creating a server. This means that in a failure cause, this list can grow without bounds.

But I don't think we need to delete these ports at all. The problem @mnaser reported could equally be characterised as: we created new ports when we didn't need to.

With the removal of networks in v1alpha7 the number and order of ports a machine has is now completely deterministic (assuming we keep spec.ports immutable). When we execute this loop we know that if CreatedPorts contains N entries we can skip the creation of the first N ports, because they have already been created. If a status Update fails we handle that because the loop does GetOrCreatePort, which will return any existing port.

We can then defer port cleanup to DeleteInstance. We are still going to have to reconstruct and iterate over expected ports as we do here because the status Update with CreatedPorts may have failed, but by only doing the resource creation once we should be able to reliably do the cleanup only once too.